flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -356,6 +356,13 @@ public Buffer getNextReceivedBuffer() {
 	@Override
 	public NotificationResult notifyBufferAvailable(Buffer buffer) {
 		NotificationResult notificationResult = NotificationResult.BUFFER_NOT_USED;
+		// Two remote channels might call this method mutually by task thread and canceller thread concurrently.
+		// To avoid deadlock issue we can check the released state to return immediately before synchronizing.
+		// See FLINK-18595 for details.
+		if (isReleased.get()) {
+			return notificationResult;
```

Comment:
Before the change, if the channel was already released while calling `notifyBufferAvailable`, it will exit directly by the following: 

```
if (isReleased.get() || bufferQueue.getAvailableBufferSize() >= numRequiredBuffers) {
	isWaitingForFloatingBuffers = false;
	return notificationResult;
}
```

So we will not leave it as true finally. 

Another option is to explicitly reset this state as false in `#releaseAllResources`, but it would also bring some additional check while checking the `isWaitingForFloatingBuffers` state to avoid misleading exception.  Even we also need to remove listeners from `LocalBufferPool` to keep consistent state. Otherwise the channel is not waiting for floating buffers, but it is still keeping as listener inside `LocalBufferPool` to bring conflict. So in the past we took the way of delay handling this state together while `notifyBufferAvailable`.

Reply:
I took a more thorough look at how `isWaitingForFloatingBuffers` is used and found that there is no change in behavior:
1. if the channel is released before `onSenderBacklog()` then `notifyBufferAvailable()` exits and ignores `isWaitingForFloatingBuffers`
1. if channel is released in the middle of `onSenderBacklog()` then `notifyBufferAvailable()` wouldn't be able to change  `isWaitingForFloatingBuffers` because of the lock - which is equivalent to this PR behavior
1. if channel is not released then behavior is the same